processor/src/trace/decoder/mod.rs

```diff
@@ -120,187 +129,593 @@ fn build_aux_col_p1<E: FieldElement<BaseField = Felt>>(
 // ================================================================================================
 
 /// Builds the execution trace of the decoder's `p3` column which describes the state of the op
 /// group table via multiset checks.
 fn build_aux_col_p3<E: FieldElement<BaseField = Felt>>(
     main_trace: &ColMatrix<Felt>,
-    trace_len: usize,
-    aux_trace_hints: &AuxTraceHints,
     alphas: &[E],
 ) -> Vec<E> {
-    // allocate memory for the column and set the starting value to ONE
-    let mut result = unsafe { uninit_vector(trace_len) };
+    let mut result_1: Vec<E> = unsafe { uninit_vector(main_trace.num_rows()) };
+    let mut result_2: Vec<E> = unsafe { uninit_vector(main_trace.num_rows()) };
+    let mut result: Vec<E> = unsafe { uninit_vector(main_trace.num_rows()) };
+
+    let main_tr = MainTrace::new(main_trace);
+
+    result_1[0] = E::ONE;
+    result_2[0] = E::ONE;
     result[0] = E::ONE;
 
-    // compute row values and their inverses for all rows which were added to the op group table
-    let (row_values, inv_row_values) =
-        build_lookup_table_row_values(aux_trace_hints.op_group_table_rows(), main_trace, alphas);
+    for i in 0..main_trace.num_rows() - 1 {
+        result_1[i] = op_group_table_inclusions(&main_tr, alphas, i);
+        result_2[i] = op_group_table_removals(&main_tr, alphas, i);
+    }
+
+    let result_2 = batch_inversion(&result_2);
+
+    for i in 0..main_trace.num_rows() - 1 {
+        result[i + 1] = result[i] * result_1[i] * result_2[i];
+    }
+
+    result
+}
 
-    // keep track of indexes into the list of op group table rows separately for inserted and
-    // removed rows
-    let mut inserted_group_idx = 0_usize;
-    let mut removed_group_idx = 0_usize;
+/// Adds a row to the block hash table.
+fn op_group_table_inclusions<E>(main_trace: &MainTrace, alphas: &[E], i: usize) -> E
+where
+    E: FieldElement<BaseField = Felt>,
+{
+    let op_code_felt = main_trace.get_op_code(i);
+    let op_code = op_code_felt.as_int() as u8;
+
+    match op_code {
+        SPAN | RESPAN => main_trace.get_op_group_table_inclusion_multiplicand(i, alphas),
+        _ => E::ONE,
+    }
+}
+
+/// Removes a row from the block hash table.
+fn op_group_table_removals<E>(main_trace: &MainTrace, alphas: &[E], i: usize) -> E
+where
+    E: FieldElement<BaseField = Felt>,
+{
+    let delete_group_flag = main_trace.get_delta_group_count(i) * main_trace.get_is_in_span(i);
+
+    if delete_group_flag == ONE {
+        main_trace.get_op_group_table_removal_multiplicand(i, alphas)
+    } else {
+        E::ONE
+    }
+}
 
-    // keep track of the last updated row in the running product column
-    let mut result_idx = 0_usize;
+// HELPER FUNCTIONS
+// ================================================================================================
 
-    for (clk, update) in aux_trace_hints.op_group_table_hints() {
-        let clk = *clk as usize;
+struct MainTrace<'a> {
+    columns: &'a ColMatrix<Felt>,
+}
```

Comment:
Eventually, we'd probably want to combine this with the other `MainTrace` - but I think this is fine for now.

Reply:
>  I'm actually not sure about this - but let's keep it as is for now.
I was not sure about the behavior with respect to `CALL` and `SYSCALL`, is this what you have in mind? And if so, should we leave a TODO/ create an issue?

Reply:
> Eventually, we'd probably want to combine this with the other `MainTrace` - but I think this is fine for now.

Agree, once we are done with the each part, I will consolidate all of the `MainTrace` structs and methods in one place.